xclim/indices/_agro.py

```diff
@@ -1067,17 +1069,185 @@ def _get_rain_season(pram):
+        `da + offset`
+    indexer
+        Indexing parameters to compute the indicator on a temporal subset of the data.
+        It accepts the same arguments as :py:func:`xclim.indices.generic.select_time`.
+    """
+    # "WPM" method doesn't seem to work for gamma or pearson3
+    dist_and_methods = {"gamma": ["ML", "APP"], "fisk": ["ML", "APP"]}
+    if dist not in dist_and_methods:
+        raise NotImplementedError(f"The distribution `{dist}` is not supported.")
+    if method not in dist_and_methods[dist]:
+        raise NotImplementedError(
+            f"The method `{method}` is not supported for distribution `{dist}`."
+        )
+
+    if offset:
+        with xarray.set_options(keep_attrs=True):
+            da = da + convert_units_to(offset, da, context="hydro")
+
+    da, group = _preprocess_standardized_index(da, freq, window, **indexer)
+
+    def wrap_fit(da):
+        if indexer != {} and da.isnull().all():
+            select_dims = {d: 0 if d != "time" else [0, 1] for d in da.dims}
+            fitted = fit(da[select_dims], dist, method)
+            return fitted.broadcast_like(da.isel(time=0, drop=True))
+        return fit(da, dist, method)
+
+    cal_range = (
+        da.time.min().dt.strftime("%Y-%m-%d"),
+        da.time.max().dt.strftime("%Y-%m-%d"),
+    )
+    params = da.groupby(group).map(wrap_fit)
+    params.attrs = {
+        "Calibration period": cal_range,
+        "freq": freq,
+        "window": window,
+        "scipy_dist": dist,
+        "method": method,
+        "group": group,
+        "time_indexer": indexer,
+        "units": "",
+        "offset": offset,
+    }
+    return params
+
+
+def _get_standardized_index(da, params, **indexer):
+    scipy_dist = get_dist(params.attrs["scipy_dist"])
+    group_name = params.attrs["group"].rsplit(".")[-1]
+    param_names = params.dparams.values
+
+    def wrap_cdf_ppf(da, params, group_idxs):
+        if indexer != {} and np.isnan(da).all():
+            return np.zeros_like(da) * np.NaN
+
+        spi = np.zeros_like(da)
+        # loop over groups (similar to np.groupies idea, but np.groupies can't be used with custom functions)
+        for ii, group_idx in enumerate(list(dict.fromkeys(group_idxs).keys())):
+            pars = params[ii, :]
+            indices = np.argwhere(group_idxs == group_idx)
+            vals = da[indices]
+            zeros = (vals == 0).sum(axis=0)
+            vals[vals == 0] = np.NaN
+            paramsd = {param_names[jj]: pars[jj] for jj in range(len(pars))}
+            probs_of_zero = zeros / vals.shape[0]
+            dist_probs = scipy_dist.cdf(vals, **paramsd)
+            probs = probs_of_zero + ((1 - probs_of_zero) * dist_probs)
+            spi[indices] = norm.ppf(probs)
+        return spi
+
+    group_idxs = da.time.dt.month if group_name == "month" else da.time.dt.dayofyear
+    std_index = xarray.apply_ufunc(
```

Comment:
Why can't this be a groupby.map ?

Reply:
The previous formulation of those indices did rely on a `groupby.map`, but it resulted in a very slow computation. This convoluted writing is what allowed the great speed up observed.

Reply:
There is a similar procedure ``standardized_index_fit_params``, but there, I found no improvement in using this complicated formulation over a `groupby.map`, so I kept the simpler `groupby.map` there.

Reply:
Ok. Just out of curiosity, do you understand where the speed up comes from?

Reply:
It might be for the same reason that. I had the intuition to do this because : 
1. I saw this part was the bottleneck ; 
2. I knew that `numpy.groupies` or `flox` can sometimes be faster than `groupby.map`. It's not possible to use the functions with my complex implementation, so I had to reproduce approximately what is done in these functions. 

In the end, I have indeed seen a good speedup. But why is this a faster way? Maybe this way of working avoids a costly regrouping step in the `groupby.map`, but I'm not sure. I think the answer is the same as why  `numpy.groupies` or `flox` can sometimes be faster than `groupby.map`